modules/geoipbackend/geoipbackend.cc

```diff
@@ -307,25 +313,25 @@ GeoIPBackend::~GeoIPBackend() {
 
 bool GeoIPBackend::lookup_static(const GeoIPDomain &dom, const DNSName &search, const QType &qtype, const DNSName& qdomain, const std::string &ip, GeoIPNetmask &gl, bool v6) {
   const auto& i = dom.records.find(search);
-  int cumul_probability = 0;
+  map<uint16_t,int> cumul_probabilities;
   int probability_rnd = 1+(dns_random(1000)); // setting probability=0 means it never is used
 
   if (i != dom.records.end()) { // return static value
     for(const auto& rr : i->second) {
+      if (qtype != QType::ANY && rr.qtype != qtype) continue;
+
       if (rr.has_weight) {
         gl.netmask = (v6?128:32);
-        int comp = cumul_probability;
-        cumul_probability += rr.weight;
+        int comp = cumul_probabilities[rr.qtype.getCode()];
```

Comment:
I am bit worried that QType::ANY will no longer work here.

Reply:
Do you mean in the case where someone put a ANY record in the yaml? (I would say this is not valid)

Otherwise I would actually say that this patch fixes the behaviour with the QType::ANY. Before this patch a ANY request with mixed QType records would only return 1 record (or zero due to the bug in #7051)
i.e. as I tried to demonstrate with the new regression test, if you specify *A* and *TXT* values in a record, I would expect the ANY question to return 1 record for the *A* type and another for the *TXT* record.

Reply:
I am wondering if this still makes sense, because now it can return "different" TXT-A pair if there are more than one specified.

Consider following example:
 - A 1.2.3.4, weight = 10
 - A 4.3.2.1, weight = 10
 - TXT "1-2-3-4", weight = 10
 - TXT "4-3-2-1", weight = 10

If I am reading this code correctly, it can end up returning four different kinds of results.

So, is the requirement to have corresponding records returned, or is it ok for them to behave independently of each other.

Seems that ANY query is handled correctly after all, I read it wrong the first round. But only question now remains is that is it OK that records are chosen randomly by type? If it's OK, then this works correctly.

Reply:
@Habbie @rgacogne or @cmouse who should take this decision?

Could we move forward? I'll send the backport review once this is merged.

Reply:
I think the result-set needs to contain one of TXT and one of A in this configuration to be correct.

Reply:
That's what the new implementation produces; a single random record per QTYPE.

Reply:
so can you paste what `dig any name.geo.example.com` gives?

Reply:
Yes certainly I can :), this is exactly the behavior I tried to demonstrate in the added regression test with the ANY query, but due to the randomness of the result I had to simplify the data.

So with the a setup similar to what you described:
```
    name.geo.example.com:
      - a:
          content: 1.2.3.4
          weight: 10
      - a:
          content: 4.2.3.1
          weight: 10
      - txt:
          content: 1-2-3-4
          weight: 10
      - txt:
          content: 4-2-3-1
          weight: 10
```

Here are the results:
```
#1:
name.geo.example.com.   30      IN      TXT     "4-2-3-1"
name.geo.example.com.   30      IN      A       4.2.3.1
#2:
name.geo.example.com.   30      IN      TXT     "4-2-3-1"
name.geo.example.com.   30      IN      A       4.2.3.1
#3:
name.geo.example.com.   30      IN      TXT     "1-2-3-4"
name.geo.example.com.   30      IN      A       1.2.3.4
#4:
name.geo.example.com.   30      IN      TXT     "4-2-3-1"
name.geo.example.com.   30      IN      A       4.2.3.1
#5:
name.geo.example.com.   30      IN      TXT     "1-2-3-4"
name.geo.example.com.   30      IN      A       1.2.3.4
```

Interestingly the pairs are preserved, but I'm not sure this is a feature that should be advertised ^^